File: backend/app/api/routes/status.py
from fastapi import APIRouter, HTTPException, Depends, status
from uuid import UUID

from app.models.pydantic.models import EntityStatusRecord
from app.services.status_service import StatusService
from app.services.service_factory import get_status_service

router = APIRouter()


@router.get("/", response_model=list[EntityStatusRecord] | None)
async def list_status_records(
    project_id: UUID | None = None,
    entity_id: UUID | None = None,
    status_service: StatusService = Depends(get_status_service),
):
    """List status records with optional filtering."""
    return await status_service.list_status_records(
        project_id=project_id, entity_id=entity_id
    )


@router.post("/", response_model=EntityStatusRecord)
async def create_status_record(
    status_record: EntityStatusRecord,
    status_service: StatusService = Depends(get_status_service),
):
    """Create a new status record."""
    try:
        return await status_service.create_status_record(status_record)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.get("/{record_id}", response_model=EntityStatusRecord)
async def get_status_record(
    record_id: UUID,
    status_service: StatusService = Depends(get_status_service),
):
    """Get a status record by ID."""
    record = await status_service.get_status_record(record_id)
    if not record:
        raise HTTPException(status_code=404, detail="Status record not found")
    return record


@router.put("/{record_id}", response_model=EntityStatusRecord)
async def update_status_record(
    record_id: UUID,
    status_record: EntityStatusRecord,
    status_service: StatusService = Depends(get_status_service),
):
    """Update an existing status record."""
    updated_record = await status_service.update_status_record(record_id, status_record)
    if not updated_record:
        raise HTTPException(status_code=404, detail="Status record not found")
    return updated_record


@router.delete("/{record_id}", response_model=bool)
async def delete_status_record(
    record_id: UUID,
    status_service: StatusService = Depends(get_status_service),
):
    """Delete a status record by ID."""
    deleted = await status_service.delete_status_record(record_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Status record not found")
    return deleted

File: frontend/src/services/status.ts
import api from './api';
import { EntityStatusRecord, EntityStatus } from '../types';

interface StatusCreateData {
  entity_id: string;
  project_id: string;
  status: EntityStatus;
  notes?: string;
  updated_by: string;
}

export const statusService = {
  async getStatusRecords(
    projectId?: string,
    entityId?: string
  ): Promise<{ data: EntityStatusRecord[] }> {
    let url = '/status';
    const params = [];

    if (projectId) params.push(`project_id=${projectId}`);
    if (entityId) params.push(`entity_id=${entityId}`);

    if (params.length > 0) {
      url += `?${params.join('&')}`;
    }

    return api.get<EntityStatusRecord[]>(url);
  },

  async getStatusRecord(id: string): Promise<{ data: EntityStatusRecord }> {
    return api.get<EntityStatusRecord>(`/status/${id}`);
  },

  async createStatusRecord(data: StatusCreateData): Promise<{ data: EntityStatusRecord }> {
    return api.post<EntityStatusRecord>('/status', {
      ...data,
      updated_at: new Date().toISOString(),
    });
  },

  async updateStatusRecord(
    id: string,
    data: Partial<StatusCreateData>
  ): Promise<{ data: EntityStatusRecord }> {
    return api.put<EntityStatusRecord>(`/status/${id}`, {
      ...data,
      updated_at: new Date().toISOString(),
    });
  },

  async deleteStatusRecord(id: string): Promise<{ data: boolean }> {
    return api.delete<boolean>(`/status/${id}`);
  },
};
